contracts/chain-adapters/UniversalStorageProof_Adapter.sol

```diff
@@ -0,0 +1,113 @@
+ * on the L2 where the SpokePool is deployed. Only the HubPool can store data to this contract.
+ */
+contract HubPoolStore {
+    error NotHubPool();
+
+    // Maps unique data hashes to data to relay to target.
+    mapping(bytes32 => bytes) public storedData;
+
+    // Counter to ensure that each stored data is unique.
+    uint256 private dataUuid;
+
+    address public immutable hubPool;
+
+    event StoredDataForTarget(bytes32 dataHash, address indexed target, bytes data, uint256 uuid);
+
+    modifier onlyHubPool() {
+        if (msg.sender != hubPool) {
+            revert NotHubPool();
+        }
+        _;
+    }
+
+    constructor(address _hubPool) {
+        hubPool = _hubPool;
+    }
+
+    function storeDataForTarget(address target, bytes calldata data) external onlyHubPool {
+        uint256 nonce = dataUuid++;
+        _storeData(target, data, nonce);
+    }
+
+    function storeDataForTargetWithNonce(
+        address target,
+        bytes calldata data,
+        uint256 nonce
+    ) external onlyHubPool {
+        _storeData(target, data, nonce);
+    }
+
+    function _storeData(
+        address target,
+        bytes calldata data,
+        uint256 nonce
+    ) internal {
+        bytes32 dataHash = keccak256(abi.encode(target, data, nonce));
+        if (storedData[dataHash].length > 0) {
+            // Data is already stored, do nothing.
+            return;
+        }
+        storedData[dataHash] = abi.encode(target, data);
+        emit StoredDataForTarget(dataHash, target, data, nonce);
+    }
+}
+
+/**
+ * @notice Stores data that can be relayed to L2 SpokePool using storage proof verification and light client contracts
+ * on the L2 where the SpokePool is deployed.
+ */
+contract UniversalStorageProof_Adapter is AdapterInterface {
+    HubPoolStore public immutable DATA_STORE;
+
+    error NotImplemented();
+
+    constructor(HubPoolStore _store) {
+        DATA_STORE = _store;
+    }
+
+    /**
+     * @notice Saves root bundle data in a simple storage contract that can be proven and relayed to L2.
+     * @param target Contract on the destination that will receive the message.
+     * @param message Data to send to target.
+     */
+    function relayMessage(address target, bytes calldata message) external payable override {
+        bytes4 selector = bytes4(message[:4]);
+        if (selector == SpokePoolInterface.relayRootBundle.selector) {
+            // Check if the message contains a relayRootBundle() call for the target SpokePool. If so, then
+            // store the data without a specific target in-mind and hardcode nonce.
+            // This is a gas optimization so that we only update a
+            // storage slot in the HubPoolStore once per root bundle execution, since the data passed to relayRootBundle
+            // will be the same for all chains. We are assuming therefore that the relayerRefundRoot and slowRelayRoot
```

Comment:
I think this is probably safe, but might be worth brainstorming if there's a way to keep the gas optimization without making this assumption, since it would lead to a post-liveness bundle execution failure if this doesn't hold true.

Reply:
How would you keep track of which "time" we are seeing this exact `refundRoot` x `slowRelayRoot` combination?

Reply:
I ended up adding a global variable to keep track of the relayRootBundle calldata. I think this maintains the gas optimization while also protecting against the case where the calldata is repeated intra bundles